File: git-client/src/stores/repo.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { RepoState } from '../types/git'
import { invoke } from '../utils/ipc'

export const useRepoStore = defineStore('repo', () => {
  const currentRepo = ref<RepoState | null>(null)
  const repoPath = ref<string>('')
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function openRepo(path: string) {
    loading.value = true
    error.value = null
    try {
      const state = await invoke<RepoState>('open_repo', { path })
      currentRepo.value = state
      repoPath.value = path
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function initRepo(path: string, bare: boolean) {
    loading.value = true
    try {
      const state = await invoke<RepoState>('init_repo', { path, bare })
      currentRepo.value = state
      repoPath.value = path
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function cloneRepo(url: string, path: string) {
    loading.value = true
    try {
      const state = await invoke<RepoState>('clone_repo', { url, path })
      currentRepo.value = state
      repoPath.value = path
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  return { currentRepo, repoPath, loading, error, openRepo, initRepo, cloneRepo }
})

File: git-client/src-tauri/src/commands/repo.rs
use crate::models::repo::RepoState;
use crate::services::repo_service;
use crate::utils::error::AppError;
use crate::AppState;
use tauri::State;

#[tauri::command]
pub async fn open_repo(
    state: State<'_, AppState>,
    path: String,
) -> Result<RepoState, AppError> {
    let repos = state.repos.clone();
    let path_clone = path.clone();
    tokio::task::spawn_blocking(move || {
        let repo_state = repo_service::open_repo(&path_clone)?;
        let mut manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        manager.open(&path_clone)?;
        Ok(repo_state)
    })
    .await?
}

#[tauri::command]
pub async fn init_repo(
    state: State<'_, AppState>,
    path: String,
    bare: bool,
) -> Result<RepoState, AppError> {
    let repos = state.repos.clone();
    let path_clone = path.clone();
    tokio::task::spawn_blocking(move || {
        let repo_state = repo_service::init_repo(&path_clone, bare)?;
        let mut manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        manager.open(&path_clone)?;
        Ok(repo_state)
    })
    .await?
}

#[tauri::command]
pub async fn clone_repo(
    state: State<'_, AppState>,
    url: String,
    path: String,
) -> Result<RepoState, AppError> {
    let repos = state.repos.clone();
    let path_clone = path.clone();
    tokio::task::spawn_blocking(move || {
        let repo_state = repo_service::clone_repo(&url, &path_clone)?;
        let mut manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        manager.open(&path_clone)?;
        Ok(repo_state)
    })
    .await?
}

#[tauri::command]
pub async fn close_repo(
    state: State<'_, AppState>,
    path: String,
) -> Result<(), AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let mut manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        manager.close(&path);
        Ok(())
    })
    .await?
}
